armor-write/src/main/java/com/rapid7/armor/write/writers/ColumnFileWriter.java

```diff
@@ -307,12 +309,354 @@ private void load(DataInputStream inputStream) throws IOException {
+    }
+
+    public SingleSection(ColumnFileSection sectionType, SectionBlob ba, SectionBlob ba2)
+    {
+      this.sectionType = sectionType;
+      this.sectionBlobs = new ArrayList<>();
+      this.sectionBlobs.add(ba);
+      this.sectionBlobs.add(ba2);
+    }
+
+    public List<SectionBlob> getSectionBlobs()
+    {
+      return sectionBlobs;
+    }
+
+    @Override public int totalSize()
+    {
+      return sectionBlobs.stream().map(e -> e.getLength()).collect(Collectors.summingInt(Integer::intValue));
+    }
+
+    public ColumnFileSection getSectionType() {
+      return this.sectionType;
+    }
+  }
+
+  static class SectionMapper {
+
+    private Section headerSection;
+
+    private Map<ColumnFileSection, Section> sections;
+    private List<Section> sectionsList;
+
+    public SectionMapper() {
+      sections = new HashMap<>();
+      sectionsList = new ArrayList<>();
+    }
+
+    public List<Section> getAll()
+       throws IOException
+    {
+      List<Section> result = new ArrayList<>();
+      result.add(headerSection);
+      result.add(computeTableOfContents(sectionsList));
+      result.addAll(sectionsList);
+      return result;
+    }
+
+    private Section computeTableOfContents(List<Section> sl)
+       throws IOException
+    {
+      SectionBuilder headerPortion = new SectionBuilder(ColumnFileSection.HEADER);
+      ByteArrayOutputStream os = headerPortion.getOutputStream();
+      // number of records in the table of contents. each record is 8 bytes in size
+      // considered alternative was to record in # of bytes, or *8, but was not taken
+      os.write(IOTools.toByteArray(sl.size()));
+      int offset = 0;
+      for( Section s : sl) {
+        os.write(IOTools.toByteArray(s.getSectionType().getSectionID()));
+        os.write(IOTools.toByteArray(offset));
+        offset += s.totalSize();
+      }
+      return headerPortion.build();
+    }
+
+    public void add(Section section)
+    {
+      if (section.getSectionType() == ColumnFileSection.HEADER) {
+        headerSection = section;
+      } else
+      {
+        sections.put(section.getSectionType(), section);
+        sectionsList.add(section);
+      }
+    }
+  }
+
+  public StreamProduct buildInputStreamV2(Compression compress) throws IOException {
+    int totalBytes = 0;
+
+    SectionMapper sections = new SectionMapper();
+
+    Section headerPortion = getHeaderSection(Constants.ColumnFileFormatVersion.VERSION_2);
+    sections.add(headerPortion);
+
+    Section metadataSection = writeMetadata(compress);
+    sections.add(metadataSection);
+
+    // Send entity dictionary
+    List<Path> tempPaths = new ArrayList<>();
+    //totalBytes += 8;
+    boolean success = false;
+    try {
+      sections.add(computeEntityDictionarySection(compress, tempPaths));
+      sections.add(computeValueDictionarySection(compress, tempPaths));
+      sections.add(computeEntityIndexSection(compress, tempPaths));
```

Comment:
Lets write the RowGroup first for v2 then the entity index. Lets see what happens.

Reply:
per our discussion, the code assumes that EntityIndex exists prior to reading RowGroup